text/0000-hidden-impls.md

```diff
@@ -0,0 +1,669 @@
+then there is no choice but to provide a way to control visibility in
+`#[derive(..)]` since you need to `#[derive(PartialEq, Eq)]` for the compiler
+to specify `#[structural_match]` for you.
+
+As syntax goes, `#[derive($vis Trait)]` is chosen as a simple but intuitive
+syntax. One could consider applying visibility to a group of crates,
+but it does not seem worth it to make this much terser.
+
+When it comes to custom deriving, it might be possible in the general case to
+ensure or automate that the implementation provided is of the visibility
+specified. However, in that case, it has to be added or checked on every
+implementation one invocation of a custom derive macro emits.
+Some custom derive macros don't emit trait implementations in the first place.
+However, automating this may be a good mechanism to consider.
+
+# Prior art
+[prior-art]: #prior-art
+
+[guarantees coherence]: http://blog.ezyang.com/2014/07/type-classes-confluence-coherence-global-uniqueness/
+
+As Rust's trait system (and thus implementations on types thereof) has its
+direct descendant in Haskell, we consider how that language deals with
+the relation between modules and type class instances (same as `impl` in Rust).
+While Haskell - without certain extensions enabled, like Rust,
+[guarantees coherence] of type class instances, it does not guarantee
+global uniqueness of instances. Haskell will allow you to define overlapping
+instances but won't allow you to actually use them - producing an error once
+you try to do that.
+
+[optional_doi]: https://link.springer.com/chapter/10.1007/978-3-319-45279-1_9
+[optional_pdf]: http://homepages.dcc.ufmg.br/~camarao/CT/optional-type-classes.pdf
+
+In a paper [Optional Type Classes for Haskell][optional_pdf] due to
+Ribero et. al [(DOI: 10.1007/978-3-319-45279-1_9)][optional_doi]
+modularization of instances is considered. This is however radically different
+than what is proposed in this RFC, which does *not* propose modularization of
+implementations. Fundamentally, that paper seems permit different instances
+of the same type class to coexist as long as the instance to use for a given
+expression is uniquely determinable. As previously discussed, such a mechanism
+would not be sound in Rust. The main difference between these approaches is
+that in our proposal, all implementations are considered when coherence checking
+an implementation. The effect of that is to prevent different crates and modules
+from defining implementations for a trait and a type when a hidden one exists.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Exact wordings of error messages discussed in this RFC and their error codes,
+as well as lint names and their phrasings are out of the scope of this RFC. 
+
+The following questions are in the scope of the RFC and should be resolved
+*before* the RFC is merged:
+
+1. Are there any strange interactions with possible plans for negative bounds?
```

Comment:
Not negative bounds, but you do need to consider effect of negative impls of auto-traits:

```rust
priv impl !Send for MyUnsafeType {}
```

I suggest that the visibility of a negative impl must be ≥ that of the trait, otherwise it would be an error.

Reply:
Hmm... this one is tricky... The impl above *should* be well-behaved (and not unsound)? and be interpreted as un-implementing `Send` for `MyUnsafeType` inside the module, but not elsewhere. Thus, the fact that `MyUnsafeType: Send` may be observed outside but not inside the module.

However; I can't see any use cases for this, and it will likely just lead to bad places; so your proposed rule sounds good, at least for now.